### ISSUE TITLE
Treat dates as strings during deserialization, or break API?

### ISSUE THREAD

**Issue by sdruskat** [citation-file-format/citation-file-format#202]
Recent development in the schema (moving to JSON Schema) have brought up an issue around how CFF handles dates:

- CFF is implemented in YAML, and YAML can represent data types such as `date` (as an example of `timestamp`, e.g., `date: 2021-06-02` as opposed to `date: "2021-06-02"`).
- Up until now, CFF has used the `timestamp` data type to handle dates (apart from the `issue-date` field)
- JSON cannot handle `timestamp`/`datetime`-like data types
- For validation, CFF is deserialized from YAML using the respective libraries in a programming language. For Python, for example, [`ruamel.yaml`](https://sourceforge.net/p/ruamel-yaml/) is being used.
- Libraries may be able to convert to native data types during deserialization. E.g., `ruamel.yaml` deserializes YAML `timestamp` values to Python `datetime.date`s.
- Now that we have the schema as a `schema.json`, clients duly expect to be able to use the respective JSON Schema library to validate against `schema.json`. For Python, e.g., we use [`jsonschema`](https://github.com/Julian/jsonschema).
- In JSON Schema, though, there is no data type `date`. Instead, this is solved via *formats*. I.e., dates are `"type": "string"` and `"format": "date"`.
- Now, when checking the deserialized YAML `datetime.date` value against the schema, validation fails because `datetime.date` is not a `string`!

Two ways to solve this from our side:

1. Leave the CFF definition as YAML `date`s as is. Document the potential issue, and leave this to clients to solve for their use case. For Python validation, for example, one can let `ruamel.yaml` deserialize only YAML `date` properties as strings so that `jsonschema` can process correctly, by saying
```python
yaml = ruamel.yaml.YAML(typ='safe')
yaml.constructor.yaml_constructors[u'tag:yaml.org,2002:timestamp'] = \
    yaml.constructor.yaml_constructors[u'tag:yaml.org,2002:str']
yml_data = yaml.load(fi)
```
(This is implemented for our use case in [#204](https://github.com/citation-file-format/citation-file-format/pull/204/files#diff-a56325a25122439fe3f0d61709038b4338ef4e6e6ce5842e7866e1928aa9bf33R11).)

2. Break API, and release version 2.0.0 (instead of 1.2.0) with dates now being defined in the schema as strings (perhaps of JSON Schema [format `date`](https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times), and/or matching a regex pattern), not YAML `date` values.

To decide whether to do 1. or 2., I'd like to get some opinions, ideally from @jspaaks (for the CFF side), @hainesr (how would you solve this for `ruby-cff`?) and @arfon (what solution would you like to see from the perspective of "downstream clients"?).

**Comment by arfon** [comment]
> To decide whether to do 1. or 2., I'd like to get some opinions, ideally from @jspaaks (for the CFF side), @hainesr (how would you solve this for ruby-cff?) and @arfon (what solution would you like to see from the perspective of "downstream clients"?).

@paxos – do you have any thoughts here? I saw that some of your updates to `ruby-cff` in https://github.com/citation-file-format/ruby-cff/pull/32 involved some modifications to the Date/Time parsing. 

> Break API, and release version 2.0.0 (instead of 1.2.0) with dates now being defined in the schema as strings (perhaps of JSON Schema format date, and/or matching a regex pattern), not YAML date values.

Personally I think this is probably the best option. Just anecdotally (when identifying fixtures/test files for https://github.com/citation-file-format/ruby-cff/pull/32) we found that many of the CITATION files in the wild aren't actually valid anyway (i.e., won't parse) so introducing a breaking change doesn't seem like a huge deal at this stage of the project?

**Comment by hainesr** [comment]
A naïve question while I think through the implications of the above: if JSON schema is unable to parse valid YAML, should we be using JSON schema to validate CFF files?

**Comment by sdruskat** [comment]
> A naïve question while I think through the implications of the above: if JSON schema is unable to parse valid YAML, should we be using JSON schema to validate CFF files?

Thanks, Rob, good point. IMHO the answer to this is twofold:

1. Yes, because JSON Schema has a proper definition, a better feature set (e.g., type unions) and supporting libraries, and use is more widespread (although this is intuition only).
2. Yes, because "clients" will are more likely used to working with JSON Schema rather than with libraries for a "YAML schema" (which isn't properly defined).

In this case, *YAML is a superset of JSON* has come back to bite us (a bit). One alternative would be to use a YAML Schema extension of JSON Schema, but [Space Telescope](https://github.com/spacetelescope/yaml-schema-standard)'s extension has been archived, and on [this (opinionated?) website](https://json-schema-everywhere.github.io/yaml), no active developments are noted.

TL;DR: I think JSON Schema is our best option at the moment.

**Comment by sdruskat** [comment]
> @paxos – do you have any thoughts here? I saw that some of your updates to `ruby-cff` in [citation-file-format/ruby-cff#32](https://github.com/citation-file-format/ruby-cff/pull/32) involved some modifications to the Date/Time parsing.

:+1: Would be great to have your opinion as well @paxos. (And if I have previously tried to explain the PR Draft feature to a GitHub employee, I apologize :wink:.)
 
> Personally I think this is probably the best option. Just anecdotally (when identifying fixtures/test files for [citation-file-format/ruby-cff#32](https://github.com/citation-file-format/ruby-cff/pull/32)) we found that many of the CITATION files in the wild aren't actually valid anyway (i.e., won't parse) so introducing a breaking change doesn't seem like a huge deal at this stage of the project?

Thanks! And yes, that's part of the reason why we've relaxed the schema a bit here and there, e.g., downgrading fields from required to optional.

**Comment by hainesr** [comment]
Thanks Stephan, I think you've convinced me 😄 I looked at validating the YAML directly back in 2018 and it seemed flaky at best.

I think I'm still a bit confused though 😆 Can't we treat dates as strings with a particular format in JSON Schema? Do we need JSON Schema to validate the date somehow? If JSON Schema can handle `title: my title` as a string, why not `date: 2021-06-03`? I'm sure you've tried this - I'm just venting really...

I've also been playing around in Ruby:
```ruby
> puts YAML.dump('date' => Date.today)
---
date: 2021-06-03

> puts YAML.dump('date' => Date.today.to_s)
---
date: '2021-06-03'
```

So we can force YAML to quote a date by forcing the date into a string first. But how we do this automatically I'm not sure yet. I think it must be possible; just need to find where in the underlying `to_yaml` stack to intercept things.

**Comment by sdruskat** [comment]
> Thanks Stephan, I think you've convinced me smile I looked at validating the YAML directly back in 2018 and it seemed flaky at best.

Yes, my experience exactly, and I'm afraid not much has changed since then.

> I think I'm still a bit confused though laughing Can't we treat dates as strings with a particular format in JSON Schema? Do we need JSON Schema to validate the date somehow? 

Sure, we can use two things to enforce string format in JSON Schema: (1) Regex pattern matching, (2) JSON Schema's [*format* option](https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times) for strings. At the moment we're [doing both](https://github.com/citation-file-format/citation-file-format/blob/b15a0cf481bcbdccd30fe88aa9a3d6e572072e02/schema.json#L424-L428).

I guess the root issue is a decision between letting the tools do more work, or asking for user behaviour to change. We can do either of two things.

1. Leave most date values in CFF defined as YAML `Date` types. In this case, tooling that wants to provides JSON Schema-based validation must make sure that the dates are deserialized into strings, not any native `date` format. In the case of Python using *ruamel.yaml*, this can be done as shown above in the [first comment](https://github.com/citation-file-format/citation-file-format/issues/202#issue-909881813). In the case of Ruby, your comment seems to suggest that this can also be done with `puts YAML.safe_load(YAML.dump('date' => Date.today.to_s), permitted_classes: [Date])`.  
Similarly, tooling that wants to provide CFF creation, dates will have to be serialized from whatever format they're in internally, to YAML `Date` types.
1. We can change the way that users are *allowed* to provide dates (in a way that "braks API"), so that they mustn't provide unquoted YAML `Date`s anymore, but must quote everything and this way provide strings. These will then be deserialized into strings by the respective YAML libraries and can work with JSON Schema without issues.

So the question is really, who we want to "burden" with this, the tool providers (i.e., us :laughing:) by having to deserialize into strings so that validation works), or the users (to use quotes and change existing CFF files so that quotes are used).

In fact, I think there is a third option as well.

3. Accept both YAML `Date`s and strings. This leaves more work for the tooling, i.e., depending on whether the supplied value is a string or `Date`, convert to string or don't. This may be the better option for users, but may require some more work during deserialization on the tooling side.

---

There is another (much less important) level here as well, in that if some tooling ignored the JSON Schema schema altogether, and validated YAML instead, YAML `Date`s are semantically more deterministic I guess.

**Comment by hainesr** [comment]
I suppose I am inclined to keep dates as YAML dates in the files themselves and deal with it internally if possible?

**Comment by sdruskat** [comment]
@jspaaks and me have talked about this now, and have decided to go with this option:

> In fact, I think there is a third option as well.
> 
> 1. Accept both YAML `Date`s and strings. This leaves more work for the tooling, i.e., depending on whether the supplied value is a string or `Date`, convert to string or don't. This may be the better option for users, but may require some more work during deserialization on the tooling side.

So, valid CFF date-related keys will be valid in these two forms:

```yaml
date-in-date: 2021-06-09
date-in-string: "2021-06-09"
```

These date-related fields are (still) **not** valid CFF:

```yaml
date-with-bad-format-in-string: "2021-06-99"
date-with-bad-format-in-string: "04-2021"
```

This means that tools wanting to read CFF will have to be able to treat dates as follows:

```pseudocode
# Pseudocode

cff = yaml.safe_load(CFF_file)
if cff.date instance of String:
    # Leave as is because JSON Schema will be able to validate.
    # If processed further here, make sure to validate in some way.
else if cff.date instance of Datetime.Date:
    String date_string = cff.date.to_string()
```

Or use YAML library deserialization options if available, such as 

```python
# During deserialization, read all timestamps as strings
yaml.constructor.yaml_constructors[u'tag:yaml.org,2002:timestamp'] = \
    yaml.constructor.yaml_constructors[u'tag:yaml.org,2002:str']
```

**Comment by sdruskat** [comment]
Implement for Python and add tests, merged in https://github.com/citation-file-format/citation-file-format/commit/56b5f4d20d5b966fb4bd74479e8543e22248bb0e.
